File: python/src/marketschema/adapters/transforms.py
"""Common transform functions for adapter mappings."""

from datetime import UTC, datetime, timedelta, timezone
from typing import Any

from marketschema.exceptions import TransformError

# Constants for time conversions
MS_PER_SECOND = 1000
SECONDS_PER_HOUR = 3600
JST_UTC_OFFSET_HOURS = 9


class Transforms:
    """Collection of common transform functions for adapter mappings.

    All methods are static and can be used directly as transform functions
    in ModelMapping definitions.
    """

    @staticmethod
    def to_float(value: Any) -> float:
        """Convert value to float.

        Args:
            value: Value to convert (string, int, or float)

        Returns:
            Float representation of the value

        Raises:
            TransformError: If conversion fails
        """
        try:
            return float(value)
        except (TypeError, ValueError) as e:
            raise TransformError(f"Cannot convert {value!r} to float") from e

    @staticmethod
    def to_int(value: Any) -> int:
        """Convert value to int.

        Args:
            value: Value to convert (string, float, or int)

        Returns:
            Integer representation of the value

        Raises:
            TransformError: If conversion fails
        """
        try:
            return int(value)
        except (TypeError, ValueError) as e:
            raise TransformError(f"Cannot convert {value!r} to int") from e

    @staticmethod
    def iso_timestamp(value: str) -> str:
        """Validate and normalize ISO 8601 timestamp to UTC.

        Args:
            value: ISO 8601 formatted timestamp string

        Returns:
            The validated timestamp string in ISO 8601 format (UTC)

        Raises:
            TransformError: If the timestamp is not valid ISO 8601
        """
        try:
            # Parse ISO timestamp and normalize to UTC
            dt = datetime.fromisoformat(value.replace("Z", "+00:00"))
            utc_dt = dt.astimezone(UTC)
            return utc_dt.isoformat().replace("+00:00", "Z")
        except (TypeError, ValueError) as e:
            raise TransformError(f"Invalid ISO timestamp: {value!r}") from e

    @staticmethod
    def unix_timestamp_ms(value: int | float) -> str:
        """Convert Unix timestamp in milliseconds to ISO 8601 string.

        Args:
            value: Unix timestamp in milliseconds

        Returns:
            ISO 8601 formatted timestamp string (UTC)

        Raises:
            TransformError: If conversion fails
        """
        try:
            timestamp_seconds = float(value) / MS_PER_SECOND
            dt = datetime.fromtimestamp(timestamp_seconds, tz=UTC)
            return dt.isoformat().replace("+00:00", "Z")
        except (TypeError, ValueError, OSError) as e:
            raise TransformError(f"Cannot convert {value!r} from unix ms") from e

    @staticmethod
    def unix_timestamp_sec(value: int | float) -> str:
        """Convert Unix timestamp in seconds to ISO 8601 string.

        Args:
            value: Unix timestamp in seconds

        Returns:
            ISO 8601 formatted timestamp string (UTC)

        Raises:
            TransformError: If conversion fails
        """
        try:
            dt = datetime.fromtimestamp(float(value), tz=UTC)
            return dt.isoformat().replace("+00:00", "Z")
        except (TypeError, ValueError, OSError) as e:
            raise TransformError(f"Cannot convert {value!r} from unix seconds") from e

    @staticmethod
    def side_from_string(value: str) -> str:
        """Normalize side string to lowercase buy/sell.

        Handles common variations:
        - "buy", "BUY", "Buy" -> "buy"
        - "sell", "SELL", "Sell" -> "sell"
        - "bid", "BID" -> "buy"
        - "ask", "ASK", "offer", "OFFER" -> "sell"

        Args:
            value: Side string from source data

        Returns:
            Normalized side string ("buy" or "sell")

        Raises:
            TransformError: If the value cannot be mapped to buy/sell
        """
        normalized = value.lower().strip()

        buy_values = {"buy", "bid", "b"}
        sell_values = {"sell", "ask", "offer", "s", "a"}

        if normalized in buy_values:
            return "buy"
        if normalized in sell_values:
            return "sell"

        raise TransformError(f"Cannot normalize side value: {value!r}")

    @staticmethod
    def jst_to_utc(value: str) -> str:
        """Convert JST (Japan Standard Time) timestamp to UTC ISO 8601.

        Args:
            value: ISO 8601 formatted timestamp in JST (or naive datetime assumed JST)

        Returns:
            ISO 8601 formatted timestamp string (UTC)

        Raises:
            TransformError: If conversion fails
        """
        try:
            # Parse the timestamp
            dt = datetime.fromisoformat(value.replace("Z", "+00:00"))

            # If naive (no timezone), assume JST
            if dt.tzinfo is None:
                jst_offset = timedelta(hours=JST_UTC_OFFSET_HOURS)
                dt = dt.replace(tzinfo=timezone(jst_offset))

            # Convert to UTC
            utc_dt = dt.astimezone(UTC)
            return utc_dt.isoformat().replace("+00:00", "Z")
        except (TypeError, ValueError) as e:
            raise TransformError(f"Cannot convert JST timestamp: {value!r}") from e

    @staticmethod
    def uppercase(value: str) -> str:
        """Convert string to uppercase.

        Args:
            value: String to convert

        Returns:
            Uppercase string
        """
        return str(value).upper()

    @staticmethod
    def lowercase(value: str) -> str:
        """Convert string to lowercase.

        Args:
            value: String to convert

        Returns:
            Lowercase string
        """
        return str(value).lower()


__all__ = ["Transforms"]

File: crates/marketschema-adapters/src/transforms.rs
//! Common transformation functions for data conversion.

use crate::error::TransformError;
use crate::mapping::TransformFn;
use serde_json::Value;
use std::sync::Arc;

/// Struct providing common transformation functions as associated functions.
pub struct Transforms;

impl Transforms {
    /// Converts a value to a floating-point number.
    pub fn to_float(value: &Value) -> Result<f64, TransformError> {
        match value {
            Value::Number(n) => n
                .as_f64()
                .ok_or_else(|| TransformError::new(format!("Cannot convert number to f64: {}", n))),
            Value::String(s) => s
                .parse::<f64>()
                .map_err(|_| TransformError::new(format!("Cannot parse string as f64: '{}'", s))),
            _ => Err(TransformError::new(format!(
                "Cannot convert to f64: {:?}",
                value
            ))),
        }
    }

    /// Returns a TransformFn that converts values to f64.
    pub fn to_float_fn() -> TransformFn {
        Arc::new(|v| Self::to_float(v).map(Value::from))
    }

    /// Converts a value to an integer.
    pub fn to_int(value: &Value) -> Result<i64, TransformError> {
        match value {
            Value::Number(n) => n
                .as_i64()
                .ok_or_else(|| TransformError::new(format!("Cannot convert number to i64: {}", n))),
            Value::String(s) => s
                .parse::<i64>()
                .map_err(|_| TransformError::new(format!("Cannot parse string as i64: '{}'", s))),
            _ => Err(TransformError::new(format!(
                "Cannot convert to i64: {:?}",
                value
            ))),
        }
    }

    /// Returns a TransformFn that converts values to i64.
    pub fn to_int_fn() -> TransformFn {
        Arc::new(|v| Self::to_int(v).map(Value::from))
    }

    /// Converts an ISO 8601 timestamp string to UTC normalized format.
    ///
    /// Uses RFC 3339 format with automatic sub-second precision to preserve
    /// millisecond/microsecond information from the input.
    pub fn iso_timestamp(value: &Value) -> Result<String, TransformError> {
        let s = value.as_str().ok_or_else(|| {
            TransformError::new(format!("Expected string for timestamp: {:?}", value))
        })?;

        // Parse and re-format to ensure UTC with Z suffix
        let dt = chrono::DateTime::parse_from_rfc3339(s)
            .map_err(|e| TransformError::new(format!("Invalid ISO timestamp '{}': {}", s, e)))?;

        // Use RFC 3339 format with automatic sub-second precision
        Ok(dt
            .with_timezone(&chrono::Utc)
            .to_rfc3339_opts(chrono::SecondsFormat::AutoSi, true))
    }

    /// Returns a TransformFn for ISO timestamp normalization.
    pub fn iso_timestamp_fn() -> TransformFn {
        Arc::new(|v| Self::iso_timestamp(v).map(Value::from))
    }

    /// Converts Unix milliseconds timestamp to ISO 8601 UTC string.
    pub fn unix_timestamp_ms(value: &Value) -> Result<String, TransformError> {
        let ms = Self::to_int(value)?;
        if ms < 0 {
            return Err(TransformError::new(format!(
                "Negative timestamp not allowed: {}",
                ms
            )));
        }

        const MS_PER_SECOND: i64 = 1000;
        let secs = ms / MS_PER_SECOND;
        let nsecs = ((ms % MS_PER_SECOND) * 1_000_000) as u32;

        let dt = chrono::DateTime::from_timestamp(secs, nsecs)
            .ok_or_else(|| TransformError::new(format!("Invalid Unix timestamp ms: {}", ms)))?;

        Ok(dt.format("%Y-%m-%dT%H:%M:%SZ").to_string())
    }

    /// Returns a TransformFn for Unix millisecond timestamp conversion.
    pub fn unix_timestamp_ms_fn() -> TransformFn {
        Arc::new(|v| Self::unix_timestamp_ms(v).map(Value::from))
    }

    /// Converts Unix seconds timestamp to ISO 8601 UTC string.
    pub fn unix_timestamp_sec(value: &Value) -> Result<String, TransformError> {
        let secs = Self::to_int(value)?;
        if secs < 0 {
            return Err(TransformError::new(format!(
                "Negative timestamp not allowed: {}",
                secs
            )));
        }

        let dt = chrono::DateTime::from_timestamp(secs, 0)
            .ok_or_else(|| TransformError::new(format!("Invalid Unix timestamp sec: {}", secs)))?;

        Ok(dt.format("%Y-%m-%dT%H:%M:%SZ").to_string())
    }

    /// Returns a TransformFn for Unix seconds timestamp conversion.
    pub fn unix_timestamp_sec_fn() -> TransformFn {
        Arc::new(|v| Self::unix_timestamp_sec(v).map(Value::from))
    }

    /// Converts JST (Japan Standard Time) timestamp to UTC.
    pub fn jst_to_utc(value: &Value) -> Result<String, TransformError> {
        let s = value.as_str().ok_or_else(|| {
            TransformError::new(format!("Expected string for timestamp: {:?}", value))
        })?;

        const JST_UTC_OFFSET_HOURS: i32 = 9;

        // Try parsing with timezone first
        if let Ok(dt) = chrono::DateTime::parse_from_rfc3339(s) {
            return Ok(dt
                .with_timezone(&chrono::Utc)
                .to_rfc3339_opts(chrono::SecondsFormat::AutoSi, true));
        }

        // Parse as naive datetime and assume JST
        let naive = chrono::NaiveDateTime::parse_from_str(s, "%Y-%m-%dT%H:%M:%S")
            .or_else(|_| chrono::NaiveDateTime::parse_from_str(s, "%Y-%m-%d %H:%M:%S"))
            .map_err(|e| TransformError::new(format!("Invalid datetime '{}': {}", s, e)))?;

        let jst = chrono::FixedOffset::east_opt(JST_UTC_OFFSET_HOURS * 3600).ok_or_else(|| {
            TransformError::new(format!("Invalid JST offset: {} hours", JST_UTC_OFFSET_HOURS))
        })?;
        let dt = naive
            .and_local_timezone(jst)
            .single()
            .ok_or_else(|| TransformError::new(format!("Ambiguous datetime: {}", s)))?;

        Ok(dt
            .with_timezone(&chrono::Utc)
            .to_rfc3339_opts(chrono::SecondsFormat::AutoSi, true))
    }

    /// Returns a TransformFn for JST to UTC conversion.
    pub fn jst_to_utc_fn() -> TransformFn {
        Arc::new(|v| Self::jst_to_utc(v).map(Value::from))
    }

    /// Normalizes trade side strings to "buy" or "sell".
    pub fn side_from_string(value: &Value) -> Result<String, TransformError> {
        let s = value
            .as_str()
            .ok_or_else(|| TransformError::new(format!("Expected string for side: {:?}", value)))?;

        match s.to_lowercase().as_str() {
            "buy" | "bid" | "long" => Ok("buy".to_string()),
            "sell" | "ask" | "short" => Ok("sell".to_string()),
            _ => Err(TransformError::new(format!(
                "Cannot normalize side value: '{}'",
                s
            ))),
        }
    }

    /// Returns a TransformFn for side normalization.
    pub fn side_from_string_fn() -> TransformFn {
        Arc::new(|v| Self::side_from_string(v).map(Value::from))
    }

    /// Converts a string to uppercase.
    pub fn uppercase(value: &Value) -> Result<String, TransformError> {
        let s = value
            .as_str()
            .ok_or_else(|| TransformError::new(format!("Expected string: {:?}", value)))?;
        Ok(s.to_uppercase())
    }

    /// Returns a TransformFn for uppercase conversion.
    pub fn uppercase_fn() -> TransformFn {
        Arc::new(|v| Self::uppercase(v).map(Value::from))
    }

    /// Converts a string to lowercase.
    pub fn lowercase(value: &Value) -> Result<String, TransformError> {
        let s = value
            .as_str()
            .ok_or_else(|| TransformError::new(format!("Expected string: {:?}", value)))?;
        Ok(s.to_lowercase())
    }

    /// Returns a TransformFn for lowercase conversion.
    pub fn lowercase_fn() -> TransformFn {
        Arc::new(|v| Self::lowercase(v).map(Value::from))
    }
}
